File: forge/src/main/java/com/selfdot/cobblemontrainers/forge/ForgePermissionValidator.java
package com.selfdot.cobblemontrainers.forge;

import com.cobblemon.mod.common.api.permission.Permission;
import com.selfdot.cobblemontrainers.command.permission.PermissionValidator;
import com.selfdot.cobblemontrainers.command.permission.TrainersPermissions;
import net.minecraft.command.CommandSource;
import net.minecraft.server.command.ServerCommandSource;
import net.minecraft.util.Identifier;
import net.minecraftforge.common.MinecraftForge;
import net.minecraftforge.server.permission.PermissionAPI;
import net.minecraftforge.server.permission.events.PermissionGatherEvent;
import net.minecraftforge.server.permission.nodes.PermissionNode;
import net.minecraftforge.server.permission.nodes.PermissionTypes;

import java.util.HashMap;
import java.util.Map;

public class ForgePermissionValidator implements PermissionValidator {

    private final Map<Identifier, PermissionNode<Boolean>> nodes = new HashMap<>();

    public ForgePermissionValidator() {
        MinecraftForge.EVENT_BUS.addListener(this::onPermissionGather);
    }

    private void onPermissionGather(PermissionGatherEvent.Nodes event) {
        TrainersPermissions.all().stream().map(permission -> {
            PermissionNode<Boolean> node = new PermissionNode<>(
                permission.getIdentifier(),
                PermissionTypes.BOOLEAN,
                (player, a, b) -> {
                    if (player == null) return false;
                    return player.hasPermissionLevel(permission.getLevel().getNumericalValue());
                }
            );
            nodes.put(permission.getIdentifier(), node);
            return node;
        }).forEach(event::addNodes);
    }

    @Override
    public boolean hasPermission(CommandSource source, Permission permission) {
        if (
            !nodes.containsKey(permission.getIdentifier()) ||
            !(source instanceof ServerCommandSource serverCommandSource) ||
            serverCommandSource.getPlayer() == null
        ) {
            return source.hasPermissionLevel(permission.getLevel().getNumericalValue());
        }
        return PermissionAPI.getPermission(serverCommandSource.getPlayer(), this.nodes.get(permission.getIdentifier()));
    }

}

File: forge/src/main/kotlin/com/selfdot/cobblemontrainers/forge/ForgePermissionValidator.kt
package com.selfdot.cobblemontrainers.forge

import com.cobblemon.mod.common.api.permission.Permission
import com.selfdot.cobblemontrainers.command.permission.PermissionValidator
import com.selfdot.cobblemontrainers.command.permission.TrainersPermissions
import net.minecraft.command.CommandSource
import net.minecraft.server.command.ServerCommandSource
import net.minecraft.util.Identifier
import net.minecraftforge.common.MinecraftForge
import net.minecraftforge.server.permission.PermissionAPI
import net.minecraftforge.server.permission.events.PermissionGatherEvent
import net.minecraftforge.server.permission.nodes.PermissionNode
import net.minecraftforge.server.permission.nodes.PermissionTypes

class ForgePermissionValidator : PermissionValidator {

    private val nodes = hashMapOf<Identifier, PermissionNode<Boolean>>()

    init {
        MinecraftForge.EVENT_BUS.addListener<PermissionGatherEvent.Nodes> {
            event -> event.addNodes(
                TrainersPermissions.all().map { permission ->
                    val node = PermissionNode(
                        permission.identifier,
                        PermissionTypes.BOOLEAN,
                        { player, _, _ -> player?.hasPermissionLevel(permission.level.numericalValue) == true }
                    )
                    this.nodes[permission.identifier] = node
                    node
                }
            )
        }
    }

    private fun findNode(permission: Permission) = this.nodes[permission.identifier]

    private fun extractPlayerFromSource(source: CommandSource) =
        if (source is ServerCommandSource) source.player else null

    override fun hasPermission(source: CommandSource, permission: Permission): Boolean {
        val player = this.extractPlayerFromSource(source) ?:
            return source.hasPermissionLevel(permission.level.numericalValue)
        val node = this.findNode(permission) ?: return source.hasPermissionLevel(permission.level.numericalValue)
        return PermissionAPI.getPermission(player, node)
    }

}
